src/document-fetcher.js

```diff
@@ -0,0 +1,164 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {FetchInitDef, XhrBase, assertSuccess, setupInit} from './xhr-base';
+import {isFormDataWrapper} from './form-data-wrapper';
+import {user} from './log';
+
+export class DocumentFetcher extends XhrBase {
+  /**
+   * Creates an instance of DocumentFetcher.
+   * @param {!Window} win
+   */
+  constructor(win) {
+    super(win);
+
+    /** @private {boolean} */
+    this.viewerResponded_ = false;
+
+    /** @private {!XMLHttpRequest} */
+    this.xhr_ = new XMLHttpRequest();
+  }
+
+  /**
+   * @override
+   */
+  fetchFromNetwork_(input, init) {
+    this.viewerResponded_ = false;
+
+    return this.maybeIntercept_(input, init)
+        .then(interceptorResponse => {
+          if (interceptorResponse) {
+            this.viewerResponded_ = true;
+            return interceptorResponse;
+          }
+          // After this point, both the native `fetch` and the `fetch` polyfill
+          // will expect a native `FormData` object in the `body` property, so
+          // the native `FormData` object needs to be unwrapped.
+          if (isFormDataWrapper(init.body)) {
+            init.body = init.body.getFormData();
+          }
+
+          return (this.xhrRequest_).apply(this, arguments);
```

Comment:
Nit: `return this.xhrRequest_(input, init)`.

Reply:
Done